SharpXMPP.NUnit/Compat/TcpClientExTests.cs

```diff
@@ -0,0 +1,87 @@
+﻿using System;
+using System.Net;
+using System.Net.Sockets;
+using System.Threading;
+using System.Threading.Tasks;
+using JetBrains.dotMemoryUnit;
+using NUnit.Framework;
+
+namespace SharpXMPP.Compat
+{
+    public class TcpClientExTests
+    {
+        private static int GetFreePort()
+        {
+            var host = new TcpListener(new IPEndPoint(IPAddress.Loopback, 0));
+            host.Start();
+            try
+            {
+                return ((IPEndPoint)host.LocalEndpoint).Port;
+            }
+            finally
+            {
+                host.Stop();
+            }
+        }
+
+        /// <summary>
+        /// This test checks for TcpClient leaks which are guaranteed to happen if the requests aren't properly
+        /// cancelled.
+        /// </summary>
+        /// <remarks>
+        /// Run with dotMemoryUnit to check for leaks, run without it for basic sanity check. For details see
+        /// https://www.jetbrains.com/help/dotmemory-unit/Get_Started.html#3-run-the-test
+        /// </remarks>
+        [Test, DotMemoryUnit(FailIfRunWithoutSupport = false)]
+        public void TestCancellation()
+        {
+            const int iterations = 100;
+            int cancelled = 0;
+
+            // Connecting to a free port takes enough time for the tasks to be cancelled in time.
+            var port = GetFreePort();
+            TcpListener host = null;
+            if (Environment.OSVersion.Platform == PlatformID.Unix)
+            {
+                // Since the port trick doesn't work on Linux, do this.
```

Comment:
Elaborate please

Reply:
On Linux, connection to a non-open port immediately throws "Connection refused", so I was forced to keep the port actually open.

On macOS and Windows, this just works (100/100 connections gets cancelled successfully), but on Linux, this will be closer to 50/100 in practice. If the test will go wild/become unstable, I'm going to disable it for CI.

Reply:
Can you try to use the same IPEndpoint for server and client? https://stackoverflow.com/a/16509870/1097384

Reply:
Tried. [Doesn't work](https://github.com/vitalyster/SharpXMPP/runs/3702298295).

Reply:
I'll have to explain a bit here, and why I don't think this is a problem at all.

What I'm doing here is connecting to a _free port_ (not open!), just for the connection to never be established. It helps to cancel these connections more easily than otherwise (if the port is open).

The point of test is to just catch a bunch of in-flight connections and cancel them, nothing else.

I'm not surprised this trick doesn't work that well on some TCP stacks, and I'm actually okay with opening the port for the purpose of the test.

If the port is actually open (i.e. there's an active `TcpListener` on that port), then some of the connections succeed, because I'm not fast enough to cancel them.

That's okay for the test: that's why we are doing 100 connection attempts (aside from memory leak testing, of course). At least one of them should be cancelled.

Reply:
If you are concerned with the nature of the test (it may start failing if all of 100 connections weren't cancelled in time), I may add a check `if (cancelled == 0) Assert.Inconclusive();`

This way, the test won't be considered failed in such circumstances, and will only say we cannot deduce whether the cancel functionality works.

(which isn't that big a problem per se: we are doing best effort cancellation anyway)

Reply:
I just did not understand why we need some platform conditional statements. If the "trick" did not work for some platform, then maybe no need to use that trick at all and do the same logic for all?

Reply:
Hm, makes sense. Ok, will use the same for all platforms.